java-squid/src/main/java/org/sonar/java/se/EqualRelation.java

```diff
@@ -36,43 +43,46 @@ protected String getOperand() {
   }
 
   @Override
-  protected Boolean implies(SymbolicValueRelation relation) {
-    if (hasSameOperand(relation)) {
-      return relation instanceof EqualRelation;
+  protected Boolean isImpliedBy(AbstractSymbolicValueRelation relation) {
+    if (relation instanceof NotMethodEqualsRelation) {
+      return Boolean.FALSE;
     }
-    return null;
+    return super.isImpliedBy(relation);
   }
 
   @Override
-  protected SymbolicValueRelation underHypothesis(SymbolicValueRelation relation) {
-    return relation.hypothesisWithEqual(this);
+  public Boolean isImpliedBySymbolicValueRelation(SymbolicValueRelation relation) {
+    return relation.impliesEqual();
   }
 
   @Override
-  protected SymbolicValueRelation hypothesisWithEqual(EqualRelation relation) {
-    if (v2.equals(relation.v1)) {
-      return new EqualRelation(v1, relation.v2);
-    } else if (v1.equals(relation.v2)) {
-      return new EqualRelation(v2, relation.v1);
-    } else if (v1.equals(relation.v1)) {
-      return new EqualRelation(v2, relation.v2);
-    } else if (v2.equals(relation.v2)) {
-      return new EqualRelation(v1, relation.v1);
-    }
-    return null;
+  protected Boolean impliesEqual() {
+    return Boolean.TRUE;
+  }
+
+  @Override
+  protected Boolean impliesNotEqual() {
+    return Boolean.FALSE;
   }
 
   @Override
-  protected SymbolicValueRelation hypothesisWithNotEqual(NotEqualRelation relation) {
-    if (v2.equals(relation.v1)) {
-      return new NotEqualRelation(v1, relation.v2);
-    } else if (v1.equals(relation.v2)) {
-      return new NotEqualRelation(v2, relation.v1);
-    } else if (v1.equals(relation.v1)) {
-      return new NotEqualRelation(v2, relation.v2);
-    } else if (v2.equals(relation.v2)) {
-      return new NotEqualRelation(v1, relation.v1);
+  @CheckForNull
+  protected SymbolicValueRelation combinedAfterSymbolicValueRelation(SymbolicValueRelation relation) {
+    if (relation instanceof SymbolicValueRelation) {
+      return relation.combinedWithEqual(this);
     }
     return null;
   }
+
+  @Override
+  @CheckForNull
+  protected SymbolicValueRelation combinedWithEqual(EqualRelation relation) {
+    return new EqualRelation(v1, relation.v2);
+  }
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless "instanceof" operator; it will always return "true". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1850)